src/main/java/com/google/cloud/tools/appengine/api/deploy/DeployConfiguration.java

```diff
@@ -16,32 +16,163 @@
   @Nullable
-  String getBucket();
+  public String getImageUrl() {
+    return imageUrl;
+  }
 
+  /** Google Cloud Project ID to deploy to. */
   @Nullable
-  String getImageUrl();
+  public String getProjectId() {
+    return projectId;
+  }
 
+  /** Promote the deployed version to receive all traffic. */
   @Nullable
-  Boolean getPromote();
+  public Boolean getPromote() {
+    return promote;
+  }
 
+  /** The App Engine server to use. Users typically will never set this value. */
   @Nullable
-  String getServer();
+  public String getServer() {
+    return server;
+  }
 
+  /** Stop the previous running version when deploying and promote this new version. */
   @Nullable
-  Boolean getStopPreviousVersion();
+  public Boolean getStopPreviousVersion() {
+    return stopPreviousVersion;
+  }
 
+  /** Version to deploy. */
   @Nullable
-  String getVersion();
+  public String getVersion() {
+    return version;
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public static final class Builder {
+    @Nullable private String bucket;
+    @Nullable private List<File> deployables;
+    @Nullable private String imageUrl;
+    @Nullable private String projectId;
+    @Nullable private Boolean promote;
+    @Nullable private String server;
+    @Nullable private Boolean stopPreviousVersion;
+    @Nullable private String version;
+
+    Builder() {}
+
+    public DeployConfiguration.Builder setBucket(@Nullable String bucket) {
+      this.bucket = bucket;
+      return this;
+    }
+
+    /** Set a non-null non-empty list of deployables. */
+    public DeployConfiguration.Builder setDeployables(List<File> deployables) {
+      if (deployables == null || deployables.size() == 0) {
+        throw new IllegalArgumentException("Null/empty deployables");
+      }
+      this.deployables = deployables;
+      return this;
+    }
+
+    public DeployConfiguration.Builder setImageUrl(@Nullable String imageUrl) {
+      this.imageUrl = imageUrl;
+      return this;
+    }
+
+    public DeployConfiguration.Builder setProjectId(@Nullable String projectId) {
+      this.projectId = projectId;
+      return this;
+    }
+
+    public DeployConfiguration.Builder setPromote(@Nullable Boolean promote) {
+      this.promote = promote;
+      return this;
+    }
+
+    public DeployConfiguration.Builder setServer(@Nullable String server) {
+      this.server = server;
+      return this;
+    }
+
+    public DeployConfiguration.Builder setStopPreviousVersion(
+        @Nullable Boolean stopPreviousVersion) {
+      this.stopPreviousVersion = stopPreviousVersion;
+      return this;
+    }
+
+    public DeployConfiguration.Builder setVersion(@Nullable String version) {
+      this.version = version;
+      return this;
+    }
+
+    /** Build a {@link DeployConfiguration}. */
+    public DeployConfiguration build() {
+      if (deployables == null) {
+        throw new IllegalStateException("Missing required properties: deployables");
```

Comment:
I often think if we can't accept required parameters in the `Builder()` constructor (although having the constructor 0-arg is the most usual way, I don't know). WDYT?

Reply:
I think I agree with you. I went with the style that autobuilder was generating, but there are probably some generation specifics that led to that. I'll change it.